File: backend/shell/resolve.go
package shell

import (
	"errors"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"

	"github.com/enough/enough/backend/config"
	"github.com/enough/enough/backend/enoughhome"
)

// ResolveBash returns the path to the resolved bash executable.
// It follows the resolution order specified in B1 of the handoff spec.
func ResolveBash() (string, error) {
	if runtime.GOOS != "windows" {
		// On non-Windows, look up bash on PATH or fall back to standard locations
		if p, err := exec.LookPath("bash"); err == nil {
			if fi, err := os.Stat(p); err == nil && !fi.IsDir() {
				return p, nil
			}
		}
		for _, candidate := range []string{"/usr/bin/bash", "/bin/bash"} {
			if fi, err := os.Stat(candidate); err == nil && !fi.IsDir() {
				return candidate, nil
			}
		}
		if shellEnv := os.Getenv("SHELL"); shellEnv != "" {
			if fi, err := os.Stat(shellEnv); err == nil && !fi.IsDir() {
				return shellEnv, nil
			}
		}
		return "/bin/sh", nil
	}

	// Windows resolution order
	localAppData := os.Getenv("LOCALAPPDATA")

	// 1. shell_path in ~/.enough/config.json
	if cfg, err := config.Load(); err == nil && cfg.ShellPath != "" {
		if fi, err := os.Stat(cfg.ShellPath); err == nil && !fi.IsDir() {
			return cfg.ShellPath, nil
		}
	}

	// 2. ENOUGH_GIT_BASH_PATH env
	if custom := os.Getenv("ENOUGH_GIT_BASH_PATH"); custom != "" {
		if fi, err := os.Stat(custom); err == nil && !fi.IsDir() {
			return custom, nil
		}
	}

	// 3. %LOCALAPPDATA%\enough\git\bin\bash.exe
	// 4. %LOCALAPPDATA%\enough\git\usr\bin\bash.exe
	pGitDir := enoughhome.PortableGitDir()
	candidates := []string{
		filepath.Join(pGitDir, "bin", "bash.exe"),
		filepath.Join(pGitDir, "usr", "bin", "bash.exe"),
	}
	for _, c := range candidates {
		if fi, err := os.Stat(c); err == nil && !fi.IsDir() {
			return c, nil
		}
	}

	// 5. exec.LookPath("bash") / where bash.exe (and verify file exists)
	if p, err := exec.LookPath("bash"); err == nil {
		if fi, err := os.Stat(p); err == nil && !fi.IsDir() {
			return p, nil
		}
	}

	// 6. %ProgramFiles%\Git\bin\bash.exe
	// 7. %ProgramFiles(x86)%\Git\bin\bash.exe
	// 8. %LOCALAPPDATA%\Programs\Git\bin\bash.exe
	var systemCandidates []string
	if pf := os.Getenv("ProgramFiles"); pf != "" {
		systemCandidates = append(systemCandidates, filepath.Join(pf, "Git", "bin", "bash.exe"))
	} else {
		systemCandidates = append(systemCandidates, `C:\Program Files\Git\bin\bash.exe`)
	}
	if pf86 := os.Getenv("ProgramFiles(x86)"); pf86 != "" {
		systemCandidates = append(systemCandidates, filepath.Join(pf86, "Git", "bin", "bash.exe"))
	} else {
		systemCandidates = append(systemCandidates, `C:\Program Files (x86)\Git\bin\bash.exe`)
	}
	if localAppData != "" {
		systemCandidates = append(systemCandidates, filepath.Join(localAppData, "Programs", "Git", "bin", "bash.exe"))
	}
	for _, c := range systemCandidates {
		if fi, err := os.Stat(c); err == nil && !fi.IsDir() {
			return c, nil
		}
	}

	// 9. Derive from git.exe on PATH
	if gitPath, err := exec.LookPath("git"); err == nil {
		if absGitPath, err := filepath.Abs(gitPath); err == nil {
			parent := filepath.Dir(absGitPath)
			gitRoot := filepath.Dir(parent)
			derivedCandidates := []string{
				filepath.Join(gitRoot, "bin", "bash.exe"),
				filepath.Join(gitRoot, "usr", "bin", "bash.exe"),
			}
			for _, c := range derivedCandidates {
				if fi, err := os.Stat(c); err == nil && !fi.IsDir() {
					return c, nil
				}
			}
		}
	}

	return "", errors.New("Git Bash not found. Enough requires Git for Windows on Windows.\n" +
		"Install it using the following command in PowerShell:\n" +
		"  irm https://raw.githubusercontent.com/haiderking1/Enough/main/scripts/install-windows.ps1 | iex\n" +
		"Or set the ENOUGH_GIT_BASH_PATH environment variable.")
}

File: backend/shell/resolve.ts
// PORT: backend/shell/resolve.go

import fs from "node:fs";
import path from "node:path";
import { Effect } from "effect";
import { load } from "../config/config";
import { portable_git_dir } from "../hollowhome/portable_git";

export type shell_error = {
  readonly _tag: "ShellError";
  readonly reason: string;
  readonly cause: unknown;
};

export const shell_error = (reason: string, cause: unknown): shell_error => ({
  _tag: "ShellError",
  reason,
  cause,
});

const is_file = (p: string): boolean => {
  try { return fs.statSync(p).isFile(); } catch { return false; }
};

const look_path = (name: string): string | null => {
  const paths = (process.env.PATH ?? "").split(path.delimiter);
  for (const dir of paths) {
    const candidate = path.join(dir, name);
    if (is_file(candidate)) return candidate;
    if (process.platform === "win32") {
      const exe = candidate.endsWith(".exe") ? candidate : `${candidate}.exe`;
      if (is_file(exe)) return exe;
    }
  }
  return null;
};

// ResolveBash returns the path to the resolved bash executable.
// It follows the resolution order specified in B1 of the handoff spec.
export const resolve_bash = (): Effect.Effect<string, shell_error> =>
  Effect.gen(function* () {
    if (process.platform !== "win32") {
      const found = look_path("bash");
      if (found !== null) return found;
      for (const candidate of ["/usr/bin/bash", "/bin/bash"]) {
        if (is_file(candidate)) return candidate;
      }
      const shell_env = process.env.SHELL ?? "";
      if (shell_env !== "" && is_file(shell_env)) return shell_env;
      return "/bin/sh";
    }

    const local_app_data = process.env.LOCALAPPDATA ?? "";

    const cfg_result = yield* Effect.either(load());
    if (cfg_result._tag === "Right") {
      const shell_path = cfg_result.right.shell_path ?? "";
      if (shell_path !== "" && is_file(shell_path)) return shell_path;
    }

    const custom = process.env.HOLLOW_GIT_BASH_PATH ?? "";
    if (custom !== "" && is_file(custom)) return custom;

    const p_git_dir = portable_git_dir();
    for (const c of [
      path.join(p_git_dir, "bin", "bash.exe"),
      path.join(p_git_dir, "usr", "bin", "bash.exe"),
    ]) {
      if (is_file(c)) return c;
    }

    const bash_path = look_path("bash");
    if (bash_path !== null && is_file(bash_path)) return bash_path;

    const system_candidates: string[] = [];
    const pf = process.env.ProgramFiles ?? "";
    system_candidates.push(pf !== "" ? path.join(pf, "Git", "bin", "bash.exe") : "C:\\Program Files\\Git\\bin\\bash.exe");
    const pf86 = process.env["ProgramFiles(x86)"] ?? "";
    system_candidates.push(pf86 !== "" ? path.join(pf86, "Git", "bin", "bash.exe") : "C:\\Program Files (x86)\\Git\\bin\\bash.exe");
    if (local_app_data !== "") system_candidates.push(path.join(local_app_data, "Programs", "Git", "bin", "bash.exe"));

    for (const c of system_candidates) {
      if (is_file(c)) return c;
    }

    const git_path = look_path("git");
    if (git_path !== null) {
      const abs_git_path = path.resolve(git_path);
      const parent = path.dirname(abs_git_path);
      const git_root = path.dirname(parent);
      for (const c of [path.join(git_root, "bin", "bash.exe"), path.join(git_root, "usr", "bin", "bash.exe")]) {
        if (is_file(c)) return c;
      }
    }

    return yield* Effect.fail(shell_error(
      "Git Bash not found. Hollow requires Git for Windows on Windows.\nInstall it using the following command in PowerShell:\n  irm https://raw.githubusercontent.com/haiderking1/Hollow/main/scripts/install-windows.ps1 | iex\nOr set the HOLLOW_GIT_BASH_PATH environment variable.",
      null,
    ));
  });

/*
PORT STATUS
source path: backend/shell/resolve.go
source lines: 118
draft lines: 114
confidence: high
status: phase_a_draft
todos:
  - verify Windows PATH executable lookup semantics match exec.LookPath exactly
notes:
  - ResolveBash returns (string, error), modeled as Effect.Effect<string, shell_error>.
  - Reuses config.load and hollowhome portable_git_dir ports.
*/
